File: 0006-ZigZag-Conversion/lc0006.py
# Python solution.

class Solution:
    def convert(self, s: str, numRows: int) -> str:
        if numRows == 1: return s
        ret = [''] * min(len(s), numRows)
        rowCursor = 0
        trigger = False

        for c in s:
            ret[rowCursor] += c
            rowCursor += -1 if trigger else 1
            if rowCursor == 0 or rowCursor == numRows-1:
                trigger = not trigger

        return ''.join(ret)

File: 0006-ZigZag-Conversion/lc0006.rs
// Rust solution.

impl Solution {
    pub fn convert(s: String, num_rows: i32) -> String {
        if (num_rows == 1) { return s; }
        let mut ret: Vec<String> = vec![String::new(); num_rows as usize];
        let mut row_cursor: usize = 0;
        let mut trigger = false;

        for c in s.chars() {
            ret[row_cursor].push(c);
            row_cursor = match trigger {
                false => row_cursor+1,
                true => row_cursor-1,
            };
            if (row_cursor == 0 || row_cursor as i32 == num_rows-1) {
                trigger = !trigger;
            }
        }
        ret.join("")
    }
}
